diff --git a/dlls/ntdll/unix/loader.c b/dlls/ntdll/unix/loader.c
index 2f97852..a7cac83 100644
--- a/dlls/ntdll/unix/loader.c
+++ b/dlls/ntdll/unix/loader.c
@@ -363,7 +363,11 @@ static const char *get_pe_dir( WORD machine )
     switch(machine)
     {
     case IMAGE_FILE_MACHINE_I386:  return "/i386-windows";
-    case IMAGE_FILE_MACHINE_AMD64: return "/x86_64-windows";
+#ifdef __aarch64__
+        case IMAGE_FILE_MACHINE_AMD64: return "/aarch64-windows";
+#else
+        case IMAGE_FILE_MACHINE_AMD64: return "/x86_64-windows";
+#endif
     case IMAGE_FILE_MACHINE_ARMNT: return "/arm-windows";
     case IMAGE_FILE_MACHINE_ARM64: return "/aarch64-windows";
     default: return "";
