diff --git a/dlls/winex11.drv/x11drv_main.c b/dlls/winex11.drv/x11drv_main.c
index 22a1e37..0700706 100644
--- a/dlls/winex11.drv/x11drv_main.c
+++ b/dlls/winex11.drv/x11drv_main.c
@@ -202,6 +202,9 @@ static const char * const atom_names[NB_XATOMS - FIRST_XATOM] =
     "text/rtf",
     "text/richtext",
     "text/uri-list",
+#ifdef __ANDROID__
+    "_NET_WM_HWND",
+#endif
     "GAMESCOPE_DISPLAY_EDID_PATH",
     "GAMESCOPE_XALIA_OVERLAY",
 };
@@ -756,8 +759,9 @@ static NTSTATUS x11drv_init( void *arg )
 #ifdef SONAME_LIBXCOMPOSITE
     X11DRV_XComposite_Init();
 #endif
+#ifdef HAVE_X11_EXTENSIONS_XINPUT2_H
     x11drv_xinput2_load();
-
+#endif
     XkbUseExtension( gdi_display, NULL, NULL );
     X11DRV_InitKeyboard( gdi_display );
     X11DRV_InitMouse( gdi_display );
