diff --git a/dlls/winex11.drv/event.c b/dlls/winex11.drv/event.c
index 541eee1..5f3b872 100644
--- a/dlls/winex11.drv/event.c
+++ b/dlls/winex11.drv/event.c
@@ -700,6 +700,9 @@ static void set_focus( Display *display, HWND focus, Time time )
 
     TRACE( "setting foreground window to %p\n", focus );
 
+#ifdef __ANDROID__
+    NtUserSetForegroundWindow( focus );
+#else
     if (X11DRV_HasWindowManager( "steamcompmgr" ) || !is_net_supported( x11drv_atom(_NET_ACTIVE_WINDOW) ))
     {
         NtUserSetForegroundWindow( focus );
@@ -710,6 +713,7 @@ static void set_focus( Display *display, HWND focus, Time time )
         if (!focus) focus = threadinfo.hwndActive;
         if (focus) focus = NtUserGetAncestor( focus, GA_ROOT );
     }
+#endif
 
     if ((win = X11DRV_get_whole_window( focus )))
     {
@@ -969,8 +973,12 @@ static void focus_out( Display *display , HWND hwnd )
     /* don't reset the foreground window, if the window which is
        getting the focus is a Wine window */
 
+#ifdef __ANDROID__
+    if (!is_current_process_focused())
+#else
     if ((X11DRV_HasWindowManager( "steamcompmgr" ) || !is_net_supported( x11drv_atom(_NET_ACTIVE_WINDOW) ))
             && !is_current_process_focused())
+#endif
     {
         /* Abey : 6-Oct-99. Check again if the focus out window is the
            Foreground window, because in most cases the messages sent
@@ -978,8 +986,10 @@ static void focus_out( Display *display , HWND hwnd )
            case we don't have to change the foreground window to 0 */
         if (hwnd == NtUserGetForegroundWindow())
         {
+#ifndef __ANDROID__
             TRACE( "lost focus, setting fg to desktop\n" );
             NtUserSetForegroundWindow( NtUserGetDesktopWindow() );
+#endif
         }
     }
  }
@@ -1119,9 +1129,13 @@ static BOOL X11DRV_MapNotify( HWND hwnd, XEvent *event )
 
     if (!data->managed && !data->embedded && data->desired_state.wm_state != WithdrawnState)
     {
+#ifdef __ANDROID__
+        set_input_focus( data );
+#else
         HWND hwndFocus = get_focus();
         if (hwndFocus && NtUserIsChild( hwnd, hwndFocus ))
             set_input_focus( data );
+#endif
     }
     release_win_data( data );
     return TRUE;
