diff --git a/server/thread.h b/server/thread.h
index 057ad9e..d9b0573 100644
--- a/server/thread.h
+++ b/server/thread.h
@@ -88,6 +88,7 @@ struct thread
     int                    priority;      /* priority level */
     int                    suspend;       /* suspend count */
     int                    dbg_hidden;    /* hidden from debugger */
+    int                    bypass_proc_suspend; /* will still run if the process is suspended */
     obj_handle_t           desktop;       /* desktop handle */
     int                    desktop_users; /* number of objects using the thread desktop */
     timeout_t              creation_time; /* Thread creation time */
