diff --git a/tools/makedep.c b/tools/makedep.c
index 22a7c14..63f97b9 100644
--- a/tools/makedep.c
+++ b/tools/makedep.c
@@ -3601,7 +3601,10 @@ static void output_unix_lib( struct makefile *make )
     const char *unixlib;
 
     if (make->disabled[arch]) return;
-    if (!(unixlib = get_expanded_make_variable( make, strmake( "%s_%s", archs.str[arch], "UNIXLIB" ) ))) unixlib = make->unixlib;
+    if ((unixlib = get_expanded_make_variable( make, strmake( "%s_%s", archs.str[arch], "UNIXLIB" ) )))
+        output_symlink_rule( strmake( "%s%s", arch_install_dirs[arch], unixlib ), strmake( "%s%s", arch_install_dirs[arch], make->unixlib ), 1 );
+    else
+        unixlib = make->unixlib;
 
     strarray_add( &make->all_targets[arch], unixlib );
     add_install_rule( make, make->module, arch, unixlib,
@@ -4699,7 +4702,10 @@ int main( int argc, char *argv[] )
         strarray_add( &target_flags[arch], target );
         arch_dirs[arch] = strmake( "%s-windows/", archs.str[arch] );
         arch_pe_dirs[arch] = arch_dirs[arch];
-        arch_install_dirs[arch] = strmake( "$(libdir)/wine/%s", arch_dirs[arch] );
+        if (!strcmp( archs.str[arch], "arm64ec" ))
+            arch_install_dirs[arch] = "$(libdir)/wine/aarch64-windows/";
+        else
+            arch_install_dirs[arch] = strmake( "$(libdir)/wine/%s", arch_dirs[arch] );
         strip_progs[arch] = get_expanded_arch_var( top_makefile, "STRIP", arch );
         dll_ext[arch] = "";
     }
