diff --git a/dlls/wow64/syscall.c b/dlls/wow64/syscall.c
index d10e64f..9ff5327 100644
--- a/dlls/wow64/syscall.c
+++ b/dlls/wow64/syscall.c
@@ -733,6 +733,28 @@ static HMODULE load_64bit_module( const WCHAR *name )
 }
 
 
+/**********************************************************************
+ *           wow64GetEnvironmentVariableW
+ */
+static DWORD wow64GetEnvironmentVariableW( LPCWSTR name, LPWSTR val, DWORD size )
+{
+    UNICODE_STRING us_name, us_value;
+    NTSTATUS status;
+    DWORD len;
+
+    RtlInitUnicodeString( &us_name, name );
+    us_value.Length = 0;
+    us_value.MaximumLength = (size ? size - 1 : 0) * sizeof(WCHAR);
+    us_value.Buffer = val;
+
+    status = RtlQueryEnvironmentVariable_U( NULL, &us_name, &us_value );
+    len = us_value.Length / sizeof(WCHAR);
+    if (status == STATUS_BUFFER_TOO_SMALL) return len + 1;
+    if (status) return 0;
+    if (!size) return len + 1;
+    val[len] = 0;
+    return len;
+}
 /**********************************************************************
  *           get_cpu_dll_name
  */
@@ -746,6 +768,12 @@ static const WCHAR *get_cpu_dll_name(void)
     HANDLE key;
     ULONG size;
 
+    WCHAR *cpu_dll = (WCHAR *)buffer;
+    UINT res;
+    if ((res = wow64GetEnvironmentVariableW( L"HODLL", cpu_dll, ARRAY_SIZE(buffer) )) &&
+        res < ARRAY_SIZE(buffer))
+        return cpu_dll;
+
     switch (current_machine)
     {
     case IMAGE_FILE_MACHINE_I386:
