diff --git a/dlls/shell32/shlfileop.c b/dlls/shell32/shlfileop.c
index f24c7b7..dd09120 100644
--- a/dlls/shell32/shlfileop.c
+++ b/dlls/shell32/shlfileop.c
@@ -949,6 +949,9 @@ static void file_entry_init(FILE_ENTRY *file_entry,
     size_t file_name_len = wcslen(file_name) + 1;
     const WCHAR *ptr;
 
+    file_entry->szDirectory = NULL;
+    file_entry->szFilename = NULL;
+
     file_entry->szFullPath = malloc(file_name_len * sizeof(WCHAR));
     wcscpy(file_entry->szFullPath, file_name);
 
@@ -963,6 +966,15 @@ static void file_entry_init(FILE_ENTRY *file_entry,
         file_entry->szFilename = malloc(file_name_len * sizeof(WCHAR));
         lstrcpyW(file_entry->szFilename, ptr + 1); /* Skip over backslash. */
     }
+    else
+    {
+        file_entry->szDirectory = malloc(sizeof(WCHAR));
+        file_entry->szDirectory[0] = 0;
+
+        file_name_len = wcslen(file_entry->szFullPath) + 1;
+        file_entry->szFilename = malloc(file_name_len * sizeof(WCHAR));
+        lstrcpyW(file_entry->szFilename, file_entry->szFullPath);
+    }
 
     file_entry->attributes = attributes;
     file_entry->bFromWildcard = from_wildcard;
@@ -1189,11 +1201,16 @@ static DWORD do_copy(FILE_OPERATION *op, const FILE_ENTRY *from, const FILE_ENTR
             debugstr_w(from->szFullPath), debugstr_w(to->szFullPath), op->req->fFlags, append_file_name);
 
     /* Determine target path. */
-    wcscpy(target_dir, to->szDirectory);
+    if (to->szDirectory)
+        wcscpy(target_dir, to->szDirectory);
+    else
+        target_dir[0] = 0;
+
     wcscpy(target, to->szFullPath);
     if (append_file_name)
     {
-        PathAppendW(target_dir, to->szFilename);
+        if (to->szFilename)
+            PathAppendW(target_dir, to->szFilename);
         PathAppendW(target, from->szFilename);
     }
 
@@ -1214,7 +1231,7 @@ static DWORD do_copy(FILE_OPERATION *op, const FILE_ENTRY *from, const FILE_ENTR
         return DE_DESTSUBTREE;
 
     /* Create target dir. */
-    if (!PathFileExistsW(target_dir))
+    if (target_dir[0] && !PathFileExistsW(target_dir))
         SHCreateDirectoryExW(NULL, target_dir, NULL);
 
     /* Source contains wildcard. */
@@ -2487,3 +2504,4 @@ HRESULT WINAPI IFileOperation_Constructor(IUnknown *outer, REFIID riid, void **o
 
     return hr;
 }
+
